src/hooks/useStorage.ts

```diff
@@ -0,0 +1,66 @@
+import { useCallback, useState } from 'react';
+
+type TStorageValue = string | boolean | number;
+
+interface IUseStorageProps<T extends TStorageValue> {
+  storageKey: string;
+  initialValue: T;
+  storage?: Storage;
+}
+
+/**
+ * Hook managing access to local or session storage.
+ *
+ * @param storageKey - key of value in the storage
+ * @param initialValue - when no value is stored on storageKey, storageValue is set to initialValue
+ * @param storage - localStorage or sessionStorage
+ * @returns object containing:
+ * - storageValue - value on storageKey in the storage
+ * - storeToStorage - stores the new value into the storage
+ */
+export const useStorage = <T extends TStorageValue>({
+  storageKey,
+  initialValue,
+  storage = localStorage,
+}: IUseStorageProps<T>) => {
+  const storageValueParser = useCallback(
+    (rawStorageValue: string): T => {
+      if (typeof initialValue === 'number') {
+        return Number(rawStorageValue) as T;
+      } else if (typeof initialValue === 'boolean') {
+        return (rawStorageValue.toLowerCase() === 'true') as T;
+      }
+
+      return rawStorageValue as T;
+    },
+    [initialValue]
+  );
+
+  const [storageValue, setStorageValue] = useState<T>(() => {
+    const rawStorageValue = storage.getItem(storageKey);
+    return rawStorageValue ? storageValueParser(rawStorageValue) : initialValue;
+  });
+
+  const storeToStorage = (value: T) => {
```

Comment:
`useCallback()` is not used in this case, @patrikk0123 and @DnsZhou didn't you notice some unwanted re-renders?

Reply:
I encapsulated in the useCallback now too